artiq/firmware/runtime/session.rs

```diff
@@ -322,6 +323,30 @@ fn process_host_message(io: &Io,
     Ok(())
 }
 
+fn report_error(stream: &mut Option<&mut TcpStream>) -> Result<(), Error<SchedError>> {
+    // needed so that we can clean up the flags
+    struct AsyncErrorDropGuard;
+    impl Drop for AsyncErrorDropGuard {
+        fn drop(&mut self) {
+            unsafe { crate::rtio_mgt::ASYNC_ERROR = core::default::Default::default(); }
+        }
+    }
+    // need to give it a name ("_" is forbidden), or the compiler drops it immediately
```

Comment:
You are using the wrong abstractions. Just add a function to ``rtio_mgt.rs`` that returns the seen flags and resets them.